### PR TITLE
Lessons Learned While Replacing the RedisGraph Parser

### PR DESCRIPTION
**Pull request by jeffreylovitz**

### CODE REVIEW

**Comment by jeffreylovitz** [comment]
There is no styling around quoted lines, I notice - is the theme responsible for that?

File: content/posts/replacing-the-redisgraph-parser.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+![Multi-Part Query Railroad Diagram](/multipart_query.png)
+[Cypher railroad diagram for multi-part queries](https://s3.amazonaws.com/artifacts.opencypher.org/railroad/MultiPartQuery.html)
+
+Our original grammar was effective at interpreting a sequence of clauses as a self-contained query, but it would have required a major rewrite to interpret that same sequence as one scope capable of reading from and projecting into others. libcypher-parser had already solved this problem and many more, so we started the even more ambitious rewrite of scrapping our parser and building off an open-source project.
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: change to: "Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and —above all – open source."

File: content/posts/replacing-the-redisgraph-parser.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+![Multi-Part Query Railroad Diagram](/multipart_query.png)
+[Cypher railroad diagram for multi-part queries](https://s3.amazonaws.com/artifacts.opencypher.org/railroad/MultiPartQuery.html)
+
+Our original grammar was effective at interpreting a sequence of clauses as a self-contained query, but it would have required a major rewrite to interpret that same sequence as one scope capable of reading from and projecting into others. libcypher-parser had already solved this problem and many more, so we started the even more ambitious rewrite of scrapping our parser and building off an open-source project.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: open-source should be open source (no hyphen), according to our style guide.

File: content/posts/replacing-the-redisgraph-parser.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
+
+RedisGraph is a better project for the issues opened, features requested, and contributions made by our open-source users. While integrating libcypher-parser, we've contributed code and feedback upstream, and collaboration that improves both projects is ongoing.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: no hyphen in open source

File: content/posts/replacing-the-redisgraph-parser.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+The lesson that a vibrant open-source community enriches us all is not a new one, but I always find it heartening to be reminded of it!
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: Change to: We all know that a vibrant open source community can enrich us, but I always find it heartening to be reminded.

File: content/posts/replacing-the-redisgraph-parser.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+As an aside, it is interesting to note that instead of building a LALR parser (as Lemon does), libcypher-parser uses a PEG grammar. I won't go into the implications of this here, but Guido van Rossum has been writing a [very interesting series](https://medium.com/@gvanrossum_83706/peg-parsers-7ed72462f97c) on the subject for the curious.
+
+## Paid back with interest
+From start to finish, this undertaking took about 3 months, though even now we have open PRs to better leverage the new system. We now work with an immutable (though annotatable) AST, and have radically modified our access patterns and AST translation layer.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: 3 months should be three months

File: content/posts/replacing-the-redisgraph-parser.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+To torture the metaphor of technical debt a little more, we recouped more than the principal over the course of this refactor. libcypher-parser gave us automatic support for a  number of features, ranging from syntactic sugar like implicit edges and partially-specified traversal ranges:
+```console
+MATCH (node1)-->(node2), (src)-[*2..]->(two_or_more_hops)
+```
+To more robust query validations, built-in precedence management for operators and comparisons, and handling for more edge cases than I can recall.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: Can you rephrase this? Is this trying to convey that built-in precedence management for operators and comparisons handle more robust query validations and edge cases?

File: content/posts/replacing-the-redisgraph-parser.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+## The strength of strong abstractions
+Beyond these improvements, the greatest benefit of this refactor has been a much stronger abstraction layer between the AST and the tree of operations required to act upon it.
+
+Our earlier grammar made it easy to rely directly on the AST for the construction of operations, especially in superficially simple clauses like CREATE. This approach belies the complexity of circumstances that need to be considered, however. For example, given the query:
+```console
+MATCH (p:Person {name: 'Jeffrey'})
+CREATE (p)-[:EMPLOYED_BY]->(c: Company {name: 'Redis Labs'})
+RETURN *
+```
+The CREATE operation does not exist in isolation; it relies on projections from previous operations and itself  projects data that must be returned to the user.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: Change to: "The CREATE operation does not exist in isolation—it relies on projections from previous operations and itself  projects data that must be returned to the user."

File: content/posts/replacing-the-redisgraph-parser.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+@12  45..56  > > > > > > rel type     :`CONSTRUCTS`
+@13  59..64  > > > > > node pattern   (@14)
+@14  60..63  > > > > > > identifier   `ast`
+@15  65..76  > > RETURN               projections=[@16]
+@16  72..76  > > > projection         expression=@17
+@17  72..75  > > > > identifier       `ast`
+```
+
+## Lessons learned
+I think this captures the most important things I've learned (outside of my editor window) over the course of this project:
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: Change to: "I think this captures the most important lessons I've learned (outside of my editor window) over the course of this project:"

File: content/posts/replacing-the-redisgraph-parser.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+1) Recognize when the time has come to pay back technical debt.
+
+2) Embrace what open-source gives, and give everything you can back!
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: open source

File: content/posts/replacing-the-redisgraph-parser.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+3) Strong abstraction layers prevent many headaches.
+
+4) Obey the wisdom of "Don't reinvent the wheel."
+
+5) After you inevitably fail to follow 4 (or any other axiom of programming), think twice next time!
+
+---
+
+[RedisGraph](https://github.com/redisgraph/redisgraph) and [libcypher-parser](https://github.com/cleishm/libcypher-parser) are both actively being developed and welcome contributors at any level! This is an exciting time in the world of graph databases and query languages, and there are lots of interesting challenges to find.
```
Comment (haleykim100) [inline content/posts/replacing-the-redisgraph-parser.md]: Does it make sense to change "actively being developed" to "actively developing"? If not, don't make the change.

**Comment by gkorland** [comment]
> There is no styling around quoted lines, I notice - is the theme responsible for that?

did you check the preview? https://deploy-preview-8--redis-engineering.netlify.com/posts/replacing-the-redisgraph-parser/

File: content/posts/replacing-the-redisgraph-parser.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
+---
+title: "Lessons learned while replacing the RedisGraph parser"
+date: 2019-11-22
+authors:
+  - Jeffrey Lovitz
```
Comment (gkorland) [inline content/posts/replacing-the-redisgraph-parser.md]: authors:
  - author:
      name: "Jeffrey Lovitz"
      link: "https://github.com/jeffreylovitz"

**Comment by jeffreylovitz** [comment]
> > There is no styling around quoted lines, I notice - is the theme responsible for that?
> 
> @jeffreylovitz did you check the preview? https://deploy-preview-8--redis-engineering.netlify.com/posts/replacing-the-redisgraph-parser/

@gkorland Yes, lines 14 and 20 look just like plain text rather than blockquotes in the preview for me!

File: content/posts/replacing-the-redisgraph-parser.md (hunk 11 above)
Comment (jeffreylovitz) [reply content/posts/replacing-the-redisgraph-parser.md]: @gkorland This doesn't appear to be rendering properly, can I get an example?
